src/v/wasm/wasmedge.cc

```diff
@@ -0,0 +1,692 @@
+  schema_registry_module* mod, const WasmEdgeModule& wasmedge_module) {
+    // NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+#define REG_HOST_FN(name)                                                      \
+    host_function<&schema_registry_module::name>::reg(                         \
+      wasmedge_module, mod, #name)
+    REG_HOST_FN(get_schema_definition);
+    REG_HOST_FN(get_schema_definition_len);
+    REG_HOST_FN(get_subject_schema);
+    REG_HOST_FN(get_subject_schema_len);
+    REG_HOST_FN(create_subject_schema);
+#undef REG_HOST_FN
+}
+
+class wasmedge_engine final : public engine {
+public:
+    wasmedge_engine(
+      std::string_view user_module_name,
+      std::vector<WasmEdgeModule> modules,
+      WasmEdgeStore s,
+      WasmEdgeVM vm,
+      std::unique_ptr<transform_module> transform_module,
+      std::unique_ptr<schema_registry_module> sr_module,
+      std::unique_ptr<wasi::preview1_module> wasi_module)
+      : engine()
+      , _modules(std::move(modules))
+      , _store_ctx(std::move(s))
+      , _vm_ctx(std::move(vm))
+      , _user_module_name(user_module_name)
+      , _rp_module(std::move(transform_module))
+      , _sr_module(std::move(sr_module))
+      , _wasi_module(std::move(wasi_module)) {}
+    wasmedge_engine(const wasmedge_engine&) = delete;
+    wasmedge_engine& operator=(const wasmedge_engine&) = delete;
+    wasmedge_engine(wasmedge_engine&&) = delete;
+    wasmedge_engine& operator=(wasmedge_engine&&) = delete;
+    ~wasmedge_engine() final = default;
+
+    std::string_view function_name() const final { return _user_module_name; }
+
+    uint64_t memory_usage_size_bytes() const final {
+        auto* mod = WasmEdge_VMGetActiveModule(_vm_ctx.get());
+        if (!mod) {
+            return 0;
+        }
+        uint32_t num_mems = WasmEdge_ModuleInstanceListMemoryLength(mod);
+        std::vector<WasmEdge_String> memory_names;
+        memory_names.assign(num_mems, {});
+        uint32_t returned_mems = WasmEdge_ModuleInstanceListMemory(
+          mod, memory_names.data(), memory_names.size());
+        uint32_t num_pages = 0;
+        for (uint32_t i = 0; i < returned_mems; ++i) {
+            auto* mem = WasmEdge_ModuleInstanceFindMemory(mod, memory_names[i]);
+            if (!mem) {
+                continue;
+            }
+            num_pages += WasmEdge_MemoryInstanceGetPageSize(mem);
+        }
+        constexpr uint64_t page_size_bytes = 64_KiB;
+        return num_pages * page_size_bytes;
+    }
+
+    ss::future<> start() final { return _queue.start(); }
+
+    ss::future<> initialize() final {
+        return _queue.enqueue<void>([this] { initialize_wasi(); });
+    }
+
+    ss::future<> stop() final { return _queue.stop(); }
+
+    ss::future<model::record_batch>
+    transform(model::record_batch batch, transform_probe* probe) override {
+        vlog(
+          wasm_log.trace,
+          "Transforming batch: {}",
+          batch.header().record_count);
```

Comment:
maybe we can log the whole header here ?

Reply:
Certainly 😄